File: frontend/src/geometry/constraintResolution.ts
import type { ConstraintsState, EdgeSide, LeafConstraint, SymmetryMode } from '../types/constraints'
import type { TreeState } from '../types/tree'
import { getLeaves } from './treeGeometry'
import { reflect } from './symmetry'
import { cornerPosition } from './edgePin'

export interface Point {
  x: number
  y: number
}

export interface Resolution {
  /** false = this symmetry+boundary combination is a straight contradiction
   * (e.g. book symmetry pins x=0.5, but a left/right edge pin pins x=0/1) —
   * callers must reject the action that would produce it. */
  feasible: boolean
  /** Non-null only when the leaf's OWN position is fully determined by its
   * constraints alone, independent of any free/drag variable — used for
   * both corner-picker greying and cross-leaf collision checks. */
  point: Point | null
}

const FEASIBLE_FREE: Resolution = { feasible: true, point: null }
const INFEASIBLE: Resolution = { feasible: false, point: null }

function diagonalCorner(edge: EdgeSide): Point {
  return edge === 'top' || edge === 'left' ? { x: 0, y: 0 } : { x: 1, y: 1 }
}

function onSymmetryLine(mode: SymmetryMode, p: Point): boolean {
  if (mode === 'book') return Math.abs(p.x - 0.5) < 1e-9
  if (mode === 'diagonal') return Math.abs(p.x - p.y) < 1e-9
  return true
}

/**
 * Resolves what a single leaf's own symmetry+boundary combo implies, given
 * the global symmetry mode:
 * - `pin_corner` is always fully fixed at that corner; combined with
 *   `pin_symmetry`, it additionally requires the corner lie on the symmetry
 *   line (diagonal passes through top_left/bottom_right only; book passes
 *   through none — any book+pin_corner+pin_symmetry combo is infeasible).
 * - `pin_edge` + `pin_symmetry`: book+top/bottom resolves to that edge's
 *   midpoint; book+left/right is a straight contradiction; diagonal+any
 *   edge collapses to top_left (top/left) or bottom_right (bottom/right).
 * - Anything else (a lone edge pin, a lone symmetry pin, `pair`, `none`)
 *   still has at least one free degree of freedom — feasible, unresolved.
 * Does NOT account for `pair` mirroring on its own — see
 * `collectResolvedPoints` for the partner-mirroring behavior.
 */
export function resolveLeafConstraint(mode: SymmetryMode, constraint: LeafConstraint): Resolution {
  const pinnedToSymmetry = constraint.symmetry.kind === 'pin_symmetry'
  const boundary = constraint.boundary

  if (boundary.kind === 'pin_corner') {
    const p = cornerPosition(boundary.corner)
    if (pinnedToSymmetry && !onSymmetryLine(mode, p)) return INFEASIBLE
    return { feasible: true, point: p }
  }

  if (boundary.kind === 'pin_edge') {
    if (!pinnedToSymmetry || mode === 'none') return FEASIBLE_FREE
    if (mode === 'book') {
      if (boundary.edge === 'left' || boundary.edge === 'right') return INFEASIBLE
      return { feasible: true, point: { x: 0.5, y: boundary.edge === 'top' ? 0 : 1 } }
    }
    return { feasible: true, point: diagonalCorner(boundary.edge) }
  }

  return FEASIBLE_FREE
}

export interface ResolvedPointEntry {
  leafId: string
  point: Point
  /** True when this point is the *mirrored* position implied by a paired
   * partner's own pin, not one this leaf declared itself. */
  derived: boolean
}

/**
 * Every leaf whose position is fully determined — by its own constraints,
 * or by a paired partner's — for corner-picker greying and pre-commit
 * collision checks. Infeasible combinations never make it into the
 * constraints state (callers reject those before applying), so this only
 * has to walk feasible ones.
 */
export function collectResolvedPoints(tree: TreeState, constraints: ConstraintsState): ResolvedPointEntry[] {
  const out: ResolvedPointEntry[] = []
  for (const leafId of getLeaves(tree)) {
    const c = constraints.perLeaf[leafId]
    if (!c) continue
    const res = resolveLeafConstraint(constraints.symmetryMode, c)
    if (!res.feasible || !res.point) continue
    out.push({ leafId, point: res.point, derived: false })
    if (c.symmetry.kind === 'pair') {
      out.push({
        leafId: c.symmetry.pairedWith,
        point: reflect(constraints.symmetryMode, res.point),
        derived: true,
      })
    }
  }
  return out
}

function pointsEqual(a: Point, b: Point): boolean {
  return Math.abs(a.x - b.x) < 1e-9 && Math.abs(a.y - b.y) < 1e-9
}

/** The first resolved point (own or pair-derived) belonging to a leaf other
 * than `excludeLeafId` that collides with `point`, or null if none. */
export function findPointCollision(
  entries: ResolvedPointEntry[],
  point: Point,
  excludeLeafId: string,
): ResolvedPointEntry | null {
  return entries.find((e) => e.leafId !== excludeLeafId && pointsEqual(e.point, point)) ?? null
}

/** Convenience check used by both constraint validation and the corner-pin
 * picker's greyed-out state. */
export function isPointOccupied(
  tree: TreeState,
  constraints: ConstraintsState,
  point: Point,
  excludeLeafId: string,
): boolean {
  return findPointCollision(collectResolvedPoints(tree, constraints), point, excludeLeafId) != null
}

/** Any two distinct leaves whose resolved points (own or pair-derived)
 * coincide — used after speculatively building a candidate constraints
 * state to catch collisions a single pre-commit check might miss (e.g. a
 * newly-paired leaf's mirrored point landing on a third leaf's pin). */
export function findAnyCollision(
  entries: ResolvedPointEntry[],
): [ResolvedPointEntry, ResolvedPointEntry] | null {
  for (let i = 0; i < entries.length; i++) {
    for (let j = i + 1; j < entries.length; j++) {
      if (entries[i].leafId !== entries[j].leafId && pointsEqual(entries[i].point, entries[j].point)) {
        return [entries[i], entries[j]]
      }
    }
  }
  return null
}

File: backend/app/core/constraint_resolution.py
"""Feasibility/collapse rules for combining a leaf's independent symmetry
and boundary constraint slots. This is a line-for-line port of
`frontend/src/geometry/constraintResolution.ts` — keep the two in sync."""

from typing import List, NamedTuple, Optional, Tuple

from app.schemas.constraints import Constraints, LeafConstraint, SymmetryMode

Point = Tuple[float, float]


class Resolution(NamedTuple):
    """`feasible=False` means this symmetry+boundary combination is a
    straight contradiction (e.g. book symmetry pins x=0.5, but a left/right
    edge pin pins x=0/1) — callers must reject whatever action would
    produce it. `point` is non-None only when the leaf's OWN position is
    fully determined by its constraints alone."""

    feasible: bool
    point: Optional[Point]


_FEASIBLE_FREE = Resolution(True, None)
_INFEASIBLE = Resolution(False, None)


def corner_position(corner: str) -> Point:
    """The unit square is math-convention y-up (y=0 at the bottom, y=1 at
    the top) — the frontend applies a single compensating y-flip at the
    final screen-pixel conversion (see `geometry/edgePin.ts`)."""
    return {
        "top_left": (0.0, 1.0),
        "top_right": (1.0, 1.0),
        "bottom_left": (0.0, 0.0),
        "bottom_right": (1.0, 0.0),
    }[corner]


def edge_position(edge: str, s: float) -> Point:
    if edge == "left":
        return (0.0, s)
    if edge == "right":
        return (1.0, s)
    if edge == "top":
        return (s, 1.0)
    if edge == "bottom":
        return (s, 0.0)
    raise ValueError(f"unknown edge {edge}")


def reflect_across_symmetry(p: Point, mode: SymmetryMode) -> Point:
    if mode == SymmetryMode.BOOK:
        return (1.0 - p[0], p[1])
    if mode == SymmetryMode.DIAGONAL:
        return (p[1], p[0])
    return p


def mirror_edge(mode: SymmetryMode, edge: str) -> str:
    """Line-for-line port of `mirrorEdge` in geometry/symmetry.ts."""
    if mode == SymmetryMode.BOOK:
        if edge == "left":
            return "right"
        if edge == "right":
            return "left"
        return edge
    if mode == SymmetryMode.DIAGONAL:
        return {"top": "right", "right": "top", "bottom": "left", "left": "bottom"}[edge]
    return edge


def mirror_corner(mode: SymmetryMode, corner: str) -> str:
    """Line-for-line port of `mirrorCorner` in geometry/symmetry.ts."""
    if mode == SymmetryMode.BOOK:
        return {
            "top_left": "top_right",
            "top_right": "top_left",
            "bottom_left": "bottom_right",
            "bottom_right": "bottom_left",
        }[corner]
    if mode == SymmetryMode.DIAGONAL:
        if corner == "top_left":
            return "bottom_right"
        if corner == "bottom_right":
            return "top_left"
        return corner
    return corner


def _diagonal_corner(edge: str) -> Point:
    """A pin_edge + diagonal pin_symmetry combo collapses to wherever that
    edge meets the line x=y — see `diagonalCorner` in
    constraintResolution.ts for the y-up convention this assumes."""
    return (1.0, 1.0) if edge in ("top", "right") else (0.0, 0.0)


def _on_symmetry_line(mode: SymmetryMode, p: Point) -> bool:
    if mode == SymmetryMode.BOOK:
        return abs(p[0] - 0.5) < 1e-9
    if mode == SymmetryMode.DIAGONAL:
        return abs(p[0] - p[1]) < 1e-9
    return True


def _resolve_symmetry_boundary(mode: SymmetryMode, constraint: LeafConstraint) -> Resolution:
    """See `resolveLeafConstraint` in constraintResolution.ts for the full
    derivation of each case."""
    pinned_to_symmetry = constraint.symmetry.kind == "pin_symmetry"
    boundary = constraint.boundary

    if boundary.kind == "pin_corner":
        p = corner_position(boundary.corner)
        if pinned_to_symmetry and not _on_symmetry_line(mode, p):
            return _INFEASIBLE
        return Resolution(True, p)

    if boundary.kind == "pin_edge":
        if not pinned_to_symmetry or mode == SymmetryMode.NONE:
            return _FEASIBLE_FREE
        if mode == SymmetryMode.BOOK:
            if boundary.edge in ("left", "right"):
                return _INFEASIBLE
            return Resolution(True, (0.5, 0.0 if boundary.edge == "top" else 1.0))
        return Resolution(True, _diagonal_corner(boundary.edge))

    return _FEASIBLE_FREE


def resolve_leaf_constraint(mode: SymmetryMode, constraint: LeafConstraint) -> Resolution:
    """A `locked` slot overrides symmetry+boundary resolution entirely — it
    freezes whatever DOF they leave free at a snapshot value, so a locked
    leaf is always fully fixed regardless of its other two slots."""
    if constraint.locked.kind == "locked":
        return Resolution(True, (constraint.locked.point.x, constraint.locked.point.y))
    return _resolve_symmetry_boundary(mode, constraint)


def is_fully_fixed_by_symmetry_boundary(mode: SymmetryMode, constraint: LeafConstraint) -> bool:
    """Whether the leaf is already fully fixed by symmetry+boundary alone,
    ignoring any existing lock — used to decide whether locking it would
    freeze any actual remaining freedom."""
    res = _resolve_symmetry_boundary(mode, constraint)
    return res.feasible and res.point is not None


class ResolvedPointEntry(NamedTuple):
    leaf_id: str
    point: Point
    derived: bool


def collect_resolved_points(leaf_ids: List[str], constraints: Constraints) -> List[ResolvedPointEntry]:
    """Every leaf whose position is fully determined — by its own
    constraints, or by a paired partner's — for pre-commit collision
    checks. Mirrors `collectResolvedPoints` in constraintResolution.ts."""
    out: List[ResolvedPointEntry] = []
    for leaf_id in leaf_ids:
        c = constraints.per_leaf.get(leaf_id)
        if c is None:
            continue
        res = resolve_leaf_constraint(constraints.symmetry_mode, c)
        if not res.feasible or res.point is None:
            continue
        out.append(ResolvedPointEntry(leaf_id, res.point, False))
        if c.symmetry.kind == "pair":
            out.append(
                ResolvedPointEntry(
                    c.symmetry.paired_with,
                    reflect_across_symmetry(res.point, constraints.symmetry_mode),
                    True,
                )
            )
    return out


def _points_equal(a: Point, b: Point) -> bool:
    return abs(a[0] - b[0]) < 1e-9 and abs(a[1] - b[1]) < 1e-9


def find_any_collision(
    entries: List[ResolvedPointEntry],
) -> Optional[Tuple[ResolvedPointEntry, ResolvedPointEntry]]:
    for i in range(len(entries)):
        for j in range(i + 1, len(entries)):
            if entries[i].leaf_id != entries[j].leaf_id and _points_equal(entries[i].point, entries[j].point):
                return entries[i], entries[j]
    return None
